### PR TITLE
[move-lang] Added UniqueSet. Cleanup around UniqueMap

### PR DESCRIPTION
**Pull request by tnowacki**
- Added UniqueSet
- Cleaned up a lot of unnecessary clones in UniqueMap by reworking ModuleIdent and TName
- This also changed the API for iteration a bit, but a legacy cloning one was added for the cases that needed to clone anyway

## Motivation

- Some code cleanliness and reducing number of copies

## Test Plan

- cargo test

### CODE REVIEW

File: language/move-lang/src/to_bytecode/context.rs [hunk 1]
```diff
@@ -206,21 +204,21 @@ impl<'a> Context<'a> {
...
-        let ModuleIdent_ { address, name } = &ident.0.value;
+        let (address, name) = &ident.value;
         IR::ModuleName::new(format!("{}::{}", address, name))
     }
 
     fn translate_module_ident(ident: ModuleIdent) -> IR::ModuleIdent {
-        let ModuleIdent_ { address, name } = ident.0.value;
-        let name = Self::translate_module_name(name);
+        let (address, name) = ident.value;
+        let name = Self::translate_module_name_(name);
```
Comment (mengxu-fb) [inline language/move-lang/src/to_bytecode/context.rs:213]: If `translate_module_name_` is only used here, would it be good to just inline it?

File: language/move-lang/src/shared/unique_map.rs [hunk 2]
```diff
@@ -25,49 +25,50 @@ impl<K: TName, V> UniqueMap<K, V> {
...
-    pub fn add(&mut self, key: K, value: V) -> Result<(), K::Loc> {
+    pub fn add(&mut self, key: K, value: V) -> Result<(), (K, K::Loc)> {
+        if let Some(old_loc) = self.get_loc(&key) {
+            return Err((key, *old_loc));
+        }
         let (loc, key_) = key.drop_loc();
         let old_value = self.0.insert(key_, (loc, value));
-        if let Some((old_loc, _)) = old_value {
-            return Err(old_loc);
-        }
```
Comment (mengxu-fb) [inline language/move-lang/src/shared/unique_map.rs:33]: This should be the reason why the unit tests are failing with the following pattern:

```diff
   ┌── tests/move_check/typing/bind_duplicate_binding.move:6:23 ───
   │
 6 │         let (f, R{f}, f) = (0, R { f: 0 }, 0);
   │                       ^ Duplicate declaration for local 'f' in a given 'let'
   ·
 6 │         let (f, R{f}, f) = (0, R { f: 0 }, 0);
-  │                   - Previously declared here
+  │              - Previously declared here
   │
```

It is a matter of placing the `Previously declared here` at the first seen location or the last seen location :)

To me it seems to be a harmless one though.

Besides, test case `translated_ir_tests/move/generics/instantiation_loops/recursive_struct.move` needs an update? The last `module M3` needs to be renamed to `M4`.

Comment (tnowacki) [reply language/move-lang/src/shared/unique_map.rs:33]: Good catch and detective work! Saved me having to look this morning :) 

Updated and fixed

File: language/move-lang/src/to_bytecode/context.rs (hunk 1 above)
Comment (tnowacki) [reply language/move-lang/src/to_bytecode/context.rs:213]: Consistency mostly

**Review by mengxu-fb — APPROVED** [review]

**Pull request closed without merging**